# Snakemake workflow: ChIP-Seq

[![Snakemake](https://img.shields.io/badge/snakemake-≥6.3.0-brightgreen.svg)](https://snakemake.github.io)

A Snakemake workflow for processing ChIP-seq experiments. Input is demultiplexed FASTQ sequencing files, output is bedgraph files.

**!!!!!! PLEASE NOTE: This is under construction and currently only works out of the box for a deployment that's very specific to the Jain and Joyce Labs. The plan is to generalize this over time. !!!!!!**

## Setup

The steps below outline how to launch this pipeline on the PMACS compute cluster.

After connecting to the cluster via SSH, you should pull this repository from GitHub, which will create a new directory named after your project:

```sh
git clone https://github.com/Jain-Joyce-Labs/snakemake-chipseq.git NAME_OF_PROJECT
cd NAME_OF_PROJECT
```

Next, you need to pull in the required resources—mostly, this is a reference genome for the relevant organism, prepared for whichever alignment tool you plan to use. The pipeline currently supports two aligners: References for Bowtie2 (called "indexes") are available from [their project homepage](https://bowtie-bio.sourceforge.net/bowtie2/index.shtml), and an indexed genome for bwa can be generated by downloading a genome FASTA ([e.g. hg38](https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.40)) and [running `bwa index`](https://bio-bwa.sourceforge.net/bwa.shtml). These indices should be placed in the `resources/` directory in the repository.

Next, you need to install snakemake. The simplest way to do this is to create a virtual environment and install snakemake within it:

```sh
# we can't run snakemake from the head node, so we
# jump into an interactive session to start it:
bsub -Is bash

# wait for the session to start...

module load python/3.9.1 # default on PMACS is python 2
python -m venv venv # create the virtual environment
source venv/bin/activate # activate the environment
pip install snakemake # this line only needs to be run the first time the pipeline is installed
```

From here, the only step remaining is to launch snakemake:
```sh
snakemake --use-singularity --jobs 45 --singularity-args "--bind /home/rabdill/snakemake-chipseq/resources" --cluster 'bsub -n 16 -o {log} -M {resources.memory}000 -R "span[hosts=1]"'
```

If you want to run snakemake in headless mode (or you can't sit around staring at the interactive job until the whole pipeline finishes), create a script to run the above steps (see `submit.sh` for an example) and submit a job using something like this:

```sh
bsub -o snakemake.log -e snakemake.err bash submit.sh
```

The snakemake output you would usually monitor in the terminal will instead be written to a log file at `.snakemake/log/`.
